src/main/scala/com/databricks/spark/avro/AvroSaver.scala

```diff
@@ -0,0 +1,205 @@
+      case TimestampType => schemaBuilder.longType()
+
+      case ArrayType(elementType, _) =>
+        val builder = SchemaBuilder.builder()
+        val elementSchema = convertTypeToAvro(elementType, builder, structName)
+
+        schemaBuilder.array().items(elementSchema)
+
+      case MapType(StringType, valueType, _) =>
+        val builder = SchemaBuilder.builder()
+        val valueSchema = convertTypeToAvro(valueType, builder, structName)
+
+        schemaBuilder.map().values(valueSchema)
+
+      case StructType(_) =>
+        convertStructToAvro(dataType.asInstanceOf[StructType],
+          schemaBuilder.record(structName))
+
+      case other => sys.error(s"Unexpected type $dataType")
+    }
+  }
+
+  private def convertFieldTypeToAvro[T](dataType: DataType,
+      newFieldBuilder: SchemaBuilder.FieldTypeBuilder[T],
+      structName: String): SchemaBuilder.FieldDefault[_, _] = {
+    dataType match {
+      case ByteType => newFieldBuilder.bytesType()
+      case ShortType => newFieldBuilder.intType()
+      case IntegerType => newFieldBuilder.intType()
+      case LongType => newFieldBuilder.longType()
+      case FloatType => newFieldBuilder.floatType()
+      case DoubleType => newFieldBuilder.doubleType()
+      case DecimalType() => newFieldBuilder.stringType()
+      case StringType => newFieldBuilder.stringType()
+      case BinaryType => newFieldBuilder.bytesType()
+      case BooleanType => newFieldBuilder.booleanType()
+      case TimestampType => newFieldBuilder.longType()
+
+      case ArrayType(elementType, _) =>
+        val builder = SchemaBuilder.builder()
+        val elementSchema = convertTypeToAvro(elementType, builder, structName)
+
+        newFieldBuilder.array().items(elementSchema)
+
+      case MapType(StringType, valueType, _) =>
+        val builder = SchemaBuilder.builder()
+        val valueSchema = convertTypeToAvro(valueType, builder, structName)
+
+        newFieldBuilder.map().values(valueSchema)
+
+      case StructType(_) =>
+        convertStructToAvro(dataType.asInstanceOf[StructType],
+          newFieldBuilder.record(structName))
+
+      case other => sys.error(s"Unexpected type $dataType")
+    }
+  }
+
+  private def convertStructToAvro[T](structType: StructType, schemaBuilder: RecordBuilder[T]): T = {
+    val fieldsAssembler: FieldAssembler[T] = schemaBuilder.fields()
+    structType.fields foreach { field =>
+      val newField = fieldsAssembler.name(field.name).`type`()
+
+      if (field.nullable) {
+        convertTypeToAvro(field.dataType, newField.optional(), field.name)
+      } else {
+        convertFieldTypeToAvro(field.dataType, newField, field.name).noDefault()
+      }
+    }
+    fieldsAssembler.endRecord()
+  }
+
+  private def convertValueToAvro(item: Any, dataType: DataType, structName: String): Any = {
+    if (item == null) {
+      // Since some schema fields might be nullable, item can be null. If so, just return it
```

Comment:
Should we ensure the data type is nullable (i.e., throw exception here otherwise)? If it's not, then we may risk writing invalid data to Avro, if it's not specifically a union type. I'm not sure of the behavior in this scenario.